engines/qemu/metaservice/shell.go

```diff
@@ -0,0 +1,323 @@
+	ShellPongTimeout = ShellPingInterval * 3
+	// ShellBlockSize is the maximum number of bytes to send in a single block
+	ShellBlockSize = 16 * 1024
+	// ShellMaxMessageSize is the maximum message size we will read
+	ShellMaxMessageSize = ShellBlockSize + 4*1024
+	// ShellMaxPendingBytes is the maximum number of bytes allowed in-flight
+	ShellMaxPendingBytes = 4 * ShellBlockSize
+)
+
+type shell struct {
+	ws           *websocket.Conn
+	mWrite       sync.Mutex
+	stdin        io.WriteCloser
+	stdout       io.ReadCloser
+	stderr       io.ReadCloser
+	stdinReader  *ioext.PipeReader
+	stdoutWriter io.WriteCloser
+	stderrWriter io.WriteCloser
+	resolve      atomics.Once // Must wrap access to success/err
+	success      bool
+	err          error
+}
+
+// newShell takes a websocket and creates a shell object implementing the
+// engines.Shell interface.
+func newShell(ws *websocket.Conn) *shell {
+	stdinReader, stdin := ioext.BlockedPipe()
+	tellOut := make(chan int, 10)
+	tellErr := make(chan int, 10)
+	stdout, stdoutWriter := ioext.AsyncPipe(ShellMaxPendingBytes, tellOut)
+	stderr, stderrWriter := ioext.AsyncPipe(ShellMaxPendingBytes, tellErr)
+	stdinReader.Unblock(ShellMaxPendingBytes)
+
+	s := &shell{
+		ws:           ws,
+		stdin:        stdin,
+		stdout:       stdout,
+		stderr:       stderr,
+		stdinReader:  stdinReader,
+		stdoutWriter: stdoutWriter,
+		stderrWriter: stderrWriter,
+	}
+
+	ws.SetReadLimit(ShellMaxMessageSize)
+	ws.SetReadDeadline(time.Now().Add(ShellPongTimeout))
+	ws.SetPongHandler(s.pongHandler)
+
+	go s.writeMessages()
+	go s.readMessages()
+	go s.sendPings()
+	go s.sendAck(StreamStdout, tellOut)
+	go s.sendAck(StreamStderr, tellErr)
+
+	return s
+}
+
+func (s *shell) dispose() {
+	// Close websocket
+	s.ws.Close()
+
+	// Close all streams
+	s.stdinReader.Close()
+	s.stdoutWriter.Close()
+	s.stderrWriter.Close()
+}
+
+func (s *shell) send(message []byte) {
+	// Write message and ensure we reset the write deadline
+	s.mWrite.Lock()
+	s.ws.SetWriteDeadline(time.Now().Add(ShellWriteTimeout))
+	err := s.ws.WriteMessage(websocket.BinaryMessage, message)
+	s.mWrite.Unlock()
+
+	if err != nil {
+		s.resolve.Do(func() {
+			debug("Resolving internal error: Failed to send message, error: %s", err)
+			s.success = false
+			s.err = engines.ErrNonFatalInternalError
+			s.dispose()
+		})
+	}
+}
+
+func (s *shell) sendPings() {
+	for {
+		// Sleep for ping interval time
+		time.Sleep(ShellPingInterval)
+
+		// Write a ping message, and reset the write deadline
+		s.mWrite.Lock()
+		s.ws.SetWriteDeadline(time.Now().Add(ShellWriteTimeout))
+		err := s.ws.WriteMessage(websocket.PingMessage, []byte{})
+		s.mWrite.Unlock()
+
+		// If there is an error we resolve with internal error
+		if err != nil {
+			s.resolve.Do(func() {
+				debug("Resolving with internal-error, failed to send ping, error: %s", err)
+				s.success = false
+				s.err = engines.ErrNonFatalInternalError
+				s.dispose()
+			})
+			return
+		}
+	}
```

Comment:
Are you sure you don't want the `sendPings` method to spawn its own go routine? Seems like something that would always run as a background process, so should have the responsibility to spawn the go routine itself.